### PR TITLE
Added returning to instances

### PR DESCRIPTION
**Pull request by Lemongrass3110**
* **Addressed Issue(s)**: #5226

* **Server Mode**: Both

* **Description of Pull Request**: 
Added the possibility to return to instances when you disconnected.

Thanks to @Forte22, @attackjom and @OptimusM

### CODE REVIEW

File: src/char/char.cpp [hunk 1]
```diff
@@ -1516,8 +1518,8 @@ int char_make_new_char( struct char_session_data* sd, char* name_, int str, int
 
 	//Insert the new char entry to the database
 	if( SQL_ERROR == Sql_Query(sql_handle, "INSERT INTO `%s` (`account_id`, `char_num`, `name`, `class`, `zeny`, `status_point`, `str`, `agi`, `vit`, `int`, `dex`, `luk`, `max_hp`, `hp`,"
-		"`max_sp`, `sp`, `hair`, `hair_color`, `last_map`, `last_x`, `last_y`, `save_map`, `save_x`, `save_y`, `sex`) VALUES ("
-		"'%d', '%d', '%s', '%d', '%d',  '%d', '%d', '%d', '%d', '%d', '%d', '%d', '%u', '%u', '%u', '%u', '%d', '%d', '%s', '%d', '%d', '%s', '%d', '%d', '%c')",
+		"`max_sp`, `sp`, `hair`, `hair_color`, `last_map`, `last_x`, `last_y`, `save_map`, `save_x`, `save_y`, `sex`, `last_instanceid`) VALUES ("
+		"'%d', '%d', '%s', '%d', '%d',  '%d', '%d', '%d', '%d', '%d', '%d', '%d', '%u', '%u', '%u', '%u', '%d', '%d', '%s', '%d', '%d', '%s', '%d', '%d', '%c', '0')",
```
Comment (vstumpf) [inline src/char/char.cpp:1522]: we dont have to add this here, the sql default will handle it

File: src/map/pc.cpp [hunk 2]
```diff
@@ -1890,6 +1904,73 @@ uint8 pc_isequip(map_session_data *sd,int n)
 	return ITEM_EQUIP_ACK_OK;
 }
 
+bool pc_lastpoint_instance( map_session_data& sd ){
```
Comment (vstumpf) [inline src/map/pc.cpp:1907]: might as well use const ref, we don't modify sd here.

File: src/char/char.cpp (hunk 1 above)
Comment (Lemongrass3110) [reply src/char/char.cpp:1522]: Thought that as well and then @OptimusM reported to me that MySQL was complaining. I was confused as well, but able to reproduce it.

Comment (vstumpf) [reply src/char/char.cpp:1522]: I wasn't able to:
```
[Info]: Loading Char Data (2000000)
[Info]: Created char: account: 2000000, char: 150006, slot: 3, name: instanceidtest
[Info]: Char load request (150006)
```

we have other fields that don't have this issue, so why is this field so special?

File: src/map/pc.cpp [hunk 3]
```diff
@@ -1890,6 +1904,73 @@ uint8 pc_isequip(map_session_data *sd,int n)
...
+	// Check if the last point was an instance
+	if( sd.status.last_point_instanceid == 0 ){
+		// Nothing to do
+		return false;
+	}
+
+	// Check if returning to the instance is allowed in general
+	if( !battle_config.instance_allow_reconnect ){
+		// Return the player to his savepoint
+		return true;
```
Comment (aleos89) [inline src/map/pc.cpp:1917]:
```suggestion
		return false;
```

File: src/map/pc.cpp [hunk 4]
```diff
@@ -1890,6 +1904,73 @@ uint8 pc_isequip(map_session_data *sd,int n)
...
+	}
+
+	std::shared_ptr<s_instance_data> instance_data = util::umap_find( instances, sd.status.last_point_instanceid );
+
+	if( instance_data == nullptr ){
+		// Instance does not exist anymore, return the player to his savepoint
+		return true;
```
Comment (aleos89) [inline src/map/pc.cpp:1924]:
```suggestion
		return false;
```

File: src/map/pc.cpp [hunk 5]
```diff
@@ -1890,6 +1904,73 @@ uint8 pc_isequip(map_session_data *sd,int n)
...
+	}
+
+	switch( instance_data->mode ){
+		case IM_NONE:
+			// Cannot validate
+			break;
+		case IM_CHAR:
+			if( sd.status.char_id != instance_data->owner_id ){
+				// It is a character bound instance and the player is not the owner, return the player to his savepoint
+				return true;
```
Comment (aleos89) [inline src/map/pc.cpp:1934]:
```suggestion
				return false;
```

File: src/map/pc.cpp [hunk 6]
```diff
@@ -1890,6 +1904,73 @@ uint8 pc_isequip(map_session_data *sd,int n)
...
+			}
+			break;
+		case IM_PARTY:
+			if( sd.status.party_id != instance_data->owner_id ){
+				// It is a party bound instance and the player is not in the party, return the player to his savepoint
+				return true;
```
Comment (aleos89) [inline src/map/pc.cpp:1940]:
```suggestion
				return false;
```

File: src/map/pc.cpp [hunk 7]
```diff
@@ -1890,6 +1904,73 @@ uint8 pc_isequip(map_session_data *sd,int n)
...
+			}
+			break;
+		case IM_GUILD:
+			if( sd.status.guild_id != instance_data->owner_id ){
+				// It is a guild bound instance and the player is not in the guild, return the player to his savepoint
+				return true;
```
Comment (aleos89) [inline src/map/pc.cpp:1946]:
```suggestion
				return false;
```

File: src/map/pc.cpp [hunk 8]
```diff
@@ -1890,6 +1904,73 @@ uint8 pc_isequip(map_session_data *sd,int n)
...
+			}
+			break;
+		case IM_CLAN:
+			if( sd.status.clan_id != instance_data->owner_id ){
+				// It is a clan bound instance and the player is not in the clan, return the player to his savepoint
+				return true;
```
Comment (aleos89) [inline src/map/pc.cpp:1952]:
```suggestion
				return false;
```

File: src/map/pc.cpp [hunk 9]
```diff
@@ -1890,6 +1904,73 @@ uint8 pc_isequip(map_session_data *sd,int n)
...
+			}
+			break;
+	}
+
+	int16 mapid = map_mapname2mapid( sd.status.last_point.map );
+
+	int16 imapid = instance_mapid( mapid, sd.status.last_point_instanceid );
```
Comment (aleos89) [inline src/map/pc.cpp:1959]:
```suggestion
	int16 imapid = instance_mapid( mapid, sd.status.last_point_instanceid );
```

File: src/map/pc.cpp [hunk 10]
```diff
@@ -1890,6 +1904,73 @@ uint8 pc_isequip(map_session_data *sd,int n)
...
+
+	if( imapid < 0 ){
+		// Instance does not contain this map anymore, return the player to his savepoint
+		return true;
```
Comment (aleos89) [inline src/map/pc.cpp:1963]:
```suggestion
		return false;
```

File: src/map/pc.cpp [hunk 11]
```diff
@@ -1890,6 +1904,73 @@ uint8 pc_isequip(map_session_data *sd,int n)
...
+	}
+
+	// Overwrite the stored "source mapname" with the real mapname
+	instance_generate_mapname( mapid, sd.status.last_point_instanceid, sd.status.last_point.map );
+	// X/Y coordinates are already correct and do not need to be modified
+
+	// Dont warp the player back to his savepoint
+	return false;
```
Comment (aleos89) [inline src/map/pc.cpp:1971]:
```suggestion
	// Don't warp the player back to his savepoint
	return true;
```

File: src/map/pc.cpp [hunk 12]
```diff
@@ -2030,6 +2111,14 @@ bool pc_authok(map_session_data *sd, uint32 login_id2, time_t expiration_time, i
 	sd->vars_ok = false;
 	sd->vars_received = 0x0;
 
+	// Check if the player's last point was in an instance and check if he can be warped back into it
+	if( pc_lastpoint_instance( *sd ) ){
```
Comment (aleos89) [inline src/map/pc.cpp:2115]:
```suggestion
	if( !pc_lastpoint_is_instance( *sd ) ){
```

**Review by aleos89 — COMMENTED** [review]
Maybe I'm thinking oddly of the instance function name here. If I am, simply close my suggestions!

File: src/map/pc.cpp (hunk 2 above)
Comment (vstumpf) [reply src/map/pc.cpp:1907]: Or even better, make this a const method

[the remark this answers is not in the document]
Comment (vstumpf) [reply src/map/pc.cpp:1907]: +1

Comment (Lemongrass3110) [reply src/map/pc.cpp:1907]: We modify sd at the bottom of the function.

**Review by aleos89 — APPROVED** [review]

**Pull request merged**